clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StorePartitionDataReceiver.java

```diff
@@ -75,12 +155,2310 @@ public void write(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> cons
+     * completed header, is a prerequisite for the follower completion and is expected to occur shortly thereafter.
+     */
+    long producerTimeStamp =
+        Long.max(consumerRecord.getPubSubMessageTime(), consumerRecord.getValue().producerMetadata.messageTimestamp);
+    PubSubTopicPartition topicPartition =
+        new PubSubTopicPartitionImpl(storeIngestionTask.getVersionTopic(), partitionConsumptionState.getPartition());
+    sendIngestionHeartbeatToVT(
+        partitionConsumptionState,
+        topicPartition,
+        callback,
+        leaderMetadataWrapper,
+        leaderCompleteState,
+        producerTimeStamp);
+  }
+
+  private void sendIngestionHeartbeatToVT(
+      PartitionConsumptionState partitionConsumptionState,
+      PubSubTopicPartition topicPartition,
+      PubSubProducerCallback callback,
+      LeaderMetadataWrapper leaderMetadataWrapper,
+      LeaderCompleteState leaderCompleteState,
+      long originTimeStampMs) {
+    storeIngestionTask.sendIngestionHeartbeat(
+        partitionConsumptionState,
+        topicPartition,
+        callback,
+        leaderMetadataWrapper,
+        true,
+        true,
+        leaderCompleteState,
+        originTimeStampMs);
+  }
+
+  private LeaderProducerCallback createProducerCallback(
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord,
+      PartitionConsumptionState partitionConsumptionState,
+      LeaderProducedRecordContext leaderProducedRecordContext,
+      int partition,
+      String kafkaUrl,
+      long beforeProcessingRecordTimestampNs) {
+    return storeIngestionTask.isActiveActiveReplicationEnabled()
+        ? new ActiveActiveProducerCallback(
+            (ActiveActiveStoreIngestionTask) storeIngestionTask,
+            consumerRecord,
+            partitionConsumptionState,
+            leaderProducedRecordContext,
+            partition,
+            kafkaUrl,
+            beforeProcessingRecordTimestampNs)
+        : new LeaderProducerCallback(
+            (LeaderFollowerStoreIngestionTask) storeIngestionTask,
+            consumerRecord,
+            partitionConsumptionState,
+            leaderProducedRecordContext,
+            partition,
+            kafkaUrl,
+            beforeProcessingRecordTimestampNs);
+  }
+
+  /**
+   * Package private for testing purposes.
+   */
+  static Lazy<ByteBuffer> unwrapByteBufferFromOldValueProvider(
+      Lazy<ByteBufferValueRecord<ByteBuffer>> oldValueProvider) {
+    return Lazy.of(() -> {
+      ByteBufferValueRecord<ByteBuffer> bbValueRecord = oldValueProvider.get();
+      return bbValueRecord == null ? null : bbValueRecord.value();
+    });
+  }
+
+  public long getWriteTimestampFromKME(KafkaMessageEnvelope kme) {
+    if (kme.producerMetadata.logicalTimestamp >= 0) {
+      return kme.producerMetadata.logicalTimestamp;
+    } else {
+      return kme.producerMetadata.messageTimestamp;
+    }
+  }
+
+  private void validatePostOperationResultsAndRecord(
```

Comment:
Nit: Is it possible that we move some util methods out of this class and make the file size smaller?

Reply:
Or do you think it is wise to instead let the DataReceiver class sub-class by AA or not?